juniper_graphql_ws/src/lib.rs

```diff
@@ -0,0 +1,1068 @@
+    /// The error that is returned on failure. The formatted error will be used as the contents of
+    /// the "message" field sent back to the client.
+    type Error: Error;
+
+    /// The future configuration type.
+    type Future: Future<Output = Result<ConnectionConfig<CtxT>, Self::Error>> + Send + 'static;
+
+    /// Returns a future for the configuration to use.
+    fn init(self, params: Variables<S>) -> Self::Future;
+}
+
+impl<F, S, CtxT, Fut, E> Init<S, CtxT> for F
+where
+    S: ScalarValue,
+    F: FnOnce(Variables<S>) -> Fut + Unpin + 'static,
+    Fut: Future<Output = Result<ConnectionConfig<CtxT>, E>> + Send + 'static,
+    E: Error,
+{
+    type Error = E;
+    type Future = Fut;
+
+    fn init(self, params: Variables<S>) -> Fut {
+        self(params)
+    }
+}
+
+enum ConnectionState<S: Schema, I: Init<S::ScalarValue, S::Context>> {
+    /// PreInit is the state before a ConnectionInit message has been accepted.
+    PreInit { init: I, schema: S },
+    /// Active is the state after a ConnectionInit message has been accepted.
+    Active {
+        config: Arc<ConnectionConfig<S::Context>>,
+        stoppers: HashMap<String, oneshot::Sender<()>>,
+        schema: S,
+    },
+    /// Terminated is the state after a ConnectionInit message has been rejected.
+    Terminated,
+}
+
+impl<S: Schema, I: Init<S::ScalarValue, S::Context>> ConnectionState<S, I> {
+    // Each message we receive results in a stream of zero or more reactions. For example, a
+    // ConnectionTerminate message results in a one-item stream with the EndStream reaction.
+    async fn handle_message(
+        self,
+        msg: ClientMessage<S::ScalarValue>,
+    ) -> (Self, BoxStream<'static, Reaction<S>>) {
+        if let ClientMessage::ConnectionTerminate = msg {
+            return (self, Reaction::EndStream.to_stream());
+        }
+
+        match self {
+            Self::PreInit { init, schema } => match msg {
+                ClientMessage::ConnectionInit { payload } => match init.init(payload).await {
+                    Ok(config) => {
+                        let keep_alive_interval = config.keep_alive_interval;
+
+                        let mut s = stream::iter(vec![Reaction::ServerMessage(
+                            ServerMessage::ConnectionAck,
+                        )])
+                        .boxed();
+
+                        if keep_alive_interval > Duration::from_secs(0) {
+                            s = s
+                                .chain(
+                                    Reaction::ServerMessage(ServerMessage::ConnectionKeepAlive)
+                                        .to_stream(),
+                                )
+                                .boxed();
+                            s = s
+                                .chain(stream::unfold((), move |_| async move {
+                                    tokio::time::delay_for(keep_alive_interval).await;
```

Comment:
It is a huge bummer to use tokio just for a delay...AFAICT everything else is executor-agnostic. Is there a generic delay that only depends on `futures`?

Reply:
Agreed. I looked for a bit before adding the tokio dependency, but couldn't find anything. :(